packages/server/src/client/internal/open-feature-client.ts

```diff
@@ -276,22 +277,47 @@ export class OpenFeatureClient implements Client {
 
     const mergedContext = this.mergeContexts(invocationContext);
 
-    // this reference cannot change during the course of evaluation
-    // it may be used as a key in WeakMaps
-    const hookContext: Readonly<HookContext> = {
-      flagKey,
-      defaultValue,
-      flagValueType: flagType,
-      clientMetadata: this.metadata,
-      providerMetadata: this._provider.metadata,
-      context: mergedContext,
-      logger: this._logger,
-    };
+    // Create hook data instances for each hook
+    const hookDataMap = new WeakMap<Hook, DefaultHookData>();
+    for (const hook of allHooks) {
+      hookDataMap.set(hook, new DefaultHookData());
+    }
+
+    // Create hook context instances for each hook (stable object references for the entire evaluation)
+    // This ensures hooks can use WeakMaps with hookContext as keys across lifecycle methods
+    const hookContextMap = new WeakMap<Hook, HookContext>();
+    for (const hook of allHooks) {
+      const hookContext: HookContext = {
+        flagKey,
+        defaultValue,
+        flagValueType: flagType,
+        clientMetadata: this.metadata,
+        providerMetadata: this._provider.metadata,
+        context: mergedContext,
+        logger: this._logger,
+        hookData: hookDataMap.get(hook)!,
+      };
+      
+      // Make the core properties immutable while allowing context updates
+      Object.defineProperty(hookContext, 'flagKey', { writable: false, configurable: false });
```

Comment:
I don't love this part, but we need to freeze everything except the evaluation context.

Reply:
I think you can just freeze it. Freezing isn't a "deep" freeze. So you still have interior mutability unless you recursively freeze components.

```
const test = {
  a: "a",
  b: new Map()
};

console.log(test);
// Output
// { a: 'a', b: Map(0) {} }
Object.freeze(test);

// This interior mutability isn't prevented by freeze.
test.b.set("test", "a")

console.log(test)
// Output
// { a: 'a', b: Map(1) { 'test' => 'a' } }

// The freeze does prevent changing a property of the frozen object.
test.a = "b"

console.log(test)
// Output
// { a: 'a', b: Map(1) { 'test' => 'a' } }
// The freeze prevented the value of a from becoming 'b'.
```